021-bucket-folders.md

```diff
@@ -0,0 +1,223 @@
+- Migration applies existing files have `folderId = null` (root level)
+- New optional parameters don't affect existing API calls
+
+### Reliability (Tests & Benchmarks)
+
+#### Benchmarks
+
+Measure:
+- Folder creation latency
+- File listing with/without folder filter
+- Moving files between folders
+- Deleting folder with 1000+ files
+
+#### Tests (UI, Unit, E2E)
+
+**Unit tests:**
+- Create folder with/without custom ID
+- Duplicate folder name prevention
+- List folders with queries/search
+- Delete empty folder
+- Delete folder with force flag
+- File upload to folder
+- List files filtered by folder
+- Move file between folders
+
+**E2E tests:**
+- Create bucket → create folder → upload file → list by folder
+- Delete folder with files (should fail)
+- Delete folder with force (files move to root)
+- Permission inheritance from bucket
+- Search across folders
+
+**Console UI:**
+- Folder view in bucket files list
+- Create folder button
+- Drag-drop files into folders
+- Breadcrumb navigation
+- Folder delete confirmation
+
+### Documentation & Content
+
+**Docs needed:**
+1. API reference for new folder endpoints
+1. SDK examples for folder operations
+
+### Prior art
+
+[prior-art]: #prior-art
+
+**AWS S3:** Uses prefixes, not real folders. Files with `/` in name create virtual hierarchy. Works but confusing - users expect folders to be entities they can rename/list.
+
+**Google Drive API:** Folders are files with special MIME type. Can nest infinitely. Complex permission inheritance. Over-engineered for basic use cases.
+
+**Dropbox API:** Clear folder vs file distinction. Simple listing. Nesting supported but not required. Our pattern is similiar.
+
+**Firebase Storage:** Prefix-based like S3. No folder metadata. Awkward for apps that need folder operations.
+
+### Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+1. **Folder permissions:** Initially inherit from bucket. Future: independent folder permissions?
+2. **Nested folders:** Defer or include v1? Adds complexity (validation, infinite loops, path resolution).
+3. **Moving folders:** If we add nesting later, do we support moving folders between folders?
+4. **Folder metadata:** Do folders need custom metadata like files? Size limits?
+
+### Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+**Phase 2 - Nested folders:**
+- Add `parentFolderId` to folder documents
+- Validate against circular references
+- Path reconstruction for breadcrumbs
+- Max depth limit (e.g., 10 levels)
+
+**Phase 3 - Advanced features:**
+- Folder-level permissions (override bucket)
+- Bulk operations (move all files in folder)
+- Folder templates (create with predefined structure)
+- Shared folders (special permission model)
+- Folder hooks/events for automation
+
+**Phase 4 - Performance:**
+- Materialized paths for fast hierarchy queries
+- Denormalized file counts (already included)
+- Folder statistics (total size, last modified)
```

Comment:
This would actually be nice in step 1, should be simple to implement and will save data migration later.